llama_index/vector_stores/typesense.py

```diff
@@ -0,0 +1,258 @@
+        return self._collection
+
+    def _create_collection(self, num_dim: int) -> None:
+        fields = [
+            {"name": "vec", "type": "float[]", "num_dim": num_dim},
+            {"name": f"{self._text_key}", "type": "string"},
+            {"name": ".*", "type": "auto"},
+        ]
+        self._client.collections.create(
+            {"name": self._collection_name, "fields": fields}
+        )
+
+    def _create_upsert_docs(
+        self, embedding_results: List[NodeWithEmbedding]
+    ) -> List[dict]:
+        upsert_docs = []
+        for node in embedding_results:
+            doc = {
+                "id": node.id,
+                "vec": node.embedding,
+                f"{self._text_key}": node.node.text,
+                "ref_doc_id": node.ref_doc_id,
+                f"{self._metadata_key}": node_to_metadata_dict(node.node),
+            }
+            upsert_docs.append(doc)
+
+        return upsert_docs
+
+    @staticmethod
+    def _to_typesense_filter(standard_filters: MetadataFilters) -> str:
+        """Convert from standard dataclass to typesense filter dict."""
+        filters = {}
+        for filter in standard_filters.filters:
+            filters[filter.key] = filter.value
+
+        if "filter_by" in filters:
+            return str(filters["filter_by"])
+        else:
+            return ""
+
+    def add(
+        self,
+        embedding_results: List[NodeWithEmbedding],
+    ) -> List[str]:
+        """Add embedding results to index.
+
+        Args
+            embedding_results: List[NodeWithEmbedding]: list of embedding results
+
+        """
+        from typesense.collection import Collection
+        from typesense.exceptions import ObjectNotFound
+
+        docs = self._create_upsert_docs(embedding_results)
+
+        try:
+            collection = cast(Collection, self.collection)
+            collection.documents.import_(
+                docs, {"action": "upsert"}, batch_size=self._batch_size
+            )
+        except ObjectNotFound:
+            # Create the collection if it doesn't already exist
+            num_dim = len(embedding_results[0].embedding)
+            self._create_collection(num_dim)
+            collection.documents.import_(
+                docs, {"action": "upsert"}, batch_size=self._batch_size
+            )
+
+        return [result.id for result in embedding_results]
+
+    def delete(self, ref_doc_id: str, **delete_kwargs: Any) -> None:
+        """
+        Delete nodes using with ref_doc_id.
+
+        Args:
+            ref_doc_id (str): The doc_id of the document to delete.
+
+        """
+
+        try:
+            from typesense.collection import Collection
+
+            collection = cast(Collection, self.collection)
+        except ImportError:
+            raise ImportError("Typesense not found. Please run `pip install typesense`")
+
+        collection.documents.delete({"filter_by": f"ref_doc_id: {ref_doc_id}"})
```

Comment:
I'd recommend doing `f"ref_doc_id:={ref_doc_id}"}`

(extra `=` after colon, to do an exact filter, instead of a contains-word filter).

Reply:
Thanks! Fixed